### ISSUE TITLE
Allow new members to select old files

### ISSUE THREAD

**Issue by geiseri** [blenderskool/blaze#141]
In my usecase I tend to select a file for transfer and then send the "link" to people over chat or email.  Currently I use nextcloud shares for that but that lacks the ability to swarm files.  I tried [instant.io](https://github.com/webtorrent/instant.io) but its UX has some "gaps" compared to blaze. 

Currently members whom have not already be joined before the transfer has been started would not be able to access that file. In this case it would be nice to allow these newly joined members to select the old transfers are still being seeded by the current members of the room.  Since it looks like currently the transfer is stored in memory this could get expensive for the group members.  It might be feasible to retain past transfers in local storage so that new members to see those torrents on join.  The new room members can then select that torrent for download from the current group members.  I would suggest session storage but the maximum size would be too small to share anything meaningful. To keep the local storage from getting out of control some sort of LRU can be used to only keep the N most recent transfers.  One other option might be to pair the cache with the room so that when a member leaves a room the cache can be cleared.   This would allow the ability for members to join/leave instant rooms when their interest in the content changes.

**Comment by blenderskool** [comment]
@geiseri This is a very interesting feature. Some of my thoughts:
- This will only work when WebTorrent transfer is used. I wonder if we can make it work even when WebSockets transfer is used(as a fallback), although not a requirement.
- Right now the torrent is destroyed on the client after it has been transferred to all the active peers. This logic would need to be updated to allow the files to be seeded for _some more time_ before it gets destroyed. Perhaps each peer has the last 3-4 shared files being seeded, or maybe some sort of Last Recently Shared scheme.
- We'll have to come up with a good UI and UX for a new peer to choose the file they wish to download. Either we show a one-time modal or a button to choose the available files for download.

> It might be feasible to retain past transfers in local storage so that new members to see those torrents on join. The new room members can then select that torrent for download from the current group members. I would suggest session storage but the maximum size would be too small to share anything meaningful. To keep the local storage from getting out of control some sort of LRU can be used to only keep the N most recent transfers. One other option might be to pair the cache with the room so that when a member leaves a room the cache can be cleared. This would allow the ability for members to join/leave instant rooms when their interest in the content changes.

I don't fully understand these lines. Do you mean to retain the past transferred files or their metadata to local storage? We can't store binary data in local storage AFAIK and its size is quite limited to store full files.

Either way, this is a fairly advanced feature and I would keep it for a future major release in case we go ahead with implementing it 🙂

**Comment by geiseri** [comment]
@blenderskool I was thinking about this I am not sure it does make  sense without using a wider DHT to share a magnet link.  So this might be pushing a square peg in a round hole.   Feel free to close this if you want to clear out open issues.  It is still an awesome tool!

**Comment by Udayraj123** [comment]
@blenderskool have you considered using  Native File System API for this?
What if Blaze asks for permissions to access a local directory and just save their file-paths for resending them?
https://whatpwacando.today/file-system

I understand this may lead to a privacy issue, we can let users use the feature only on opt-in.

**Comment by blenderskool** [comment]
@Udayraj123 Yes, I've did take a look at file system API for reading / writing to a file during transfers but not for this particular task.
If we can store the paths(or references) to past shared files if user gives permission, then we can allow others to choose from these files and start the file sharing progress.

**Comment by Udayraj123** [comment]
It would be a really cool feature to have then! I'll ask for more details, in case I end up building something similar for OMRChecker. I am considering to build that on top of local rooms feature
